src/main/java/com/es/phoneshop/service/impl/DefaultCartService.java

```diff
@@ -83,43 +76,42 @@ public void add(Cart cart, Long productId, int quantity) throws OutOfStockExcept
 
     @Override
     public void update(Cart cart, Long productId, int quantity) throws OutOfStockException {
-        readWriteLock.writeLock().lock();
-        try {
+        synchronized (cart) {
             Product product = productDao.getProduct(productId);
             Optional<CartItem> foundCartItem = findCartItem(cart, product);
 
             if (product.getStock() < quantity) {
-                throw new OutOfStockException(product, product.getStock(), quantity);
+                throw new OutOfStockException("Out of stock, available: " + product.getStock() +
+                        ", requested: " + quantity);
             }
             if (foundCartItem.isPresent()) {
                 foundCartItem.get().setQuantity(quantity);
             } else {
                 CartItem cartItem = new CartItem(product, quantity);
                 cart.getItems().add(cartItem);
             }
-            recalculateCardTotalQuantity(cart);
-            recalculateCartTotalCost(cart);
-        } finally {
-            readWriteLock.writeLock().unlock();
         }
+        recalculateCart(cart);
     }
 
     @Override
     public void delete(Cart cart, Long productId) {
-        cart.getItems().removeIf(cartItem -> cartItem.getProduct().getId().equals(productId));
-        recalculateCardTotalQuantity(cart);
-        recalculateCartTotalCost(cart);
-    }
-
-    private void recalculateCardTotalQuantity(Cart cart) {
-        cart.setTotalQuantity(cart.getItems().stream()
-                .mapToInt(CartItem::getQuantity)
-                .sum());
+        synchronized (cart) {
+            cart.getItems().removeIf(cartItem -> cartItem.getProduct().getId().equals(productId));
+        }
+        recalculateCart(cart);
     }
 
-    private void recalculateCartTotalCost(Cart cart) {
-        cart.setTotalCost(cart.getItems().stream()
-                .map(cartItem -> cartItem.getProduct().getPrice().multiply(BigDecimal.valueOf(cartItem.getQuantity())))
-                .reduce(BigDecimal.ZERO, BigDecimal::add));
+    private void recalculateCart(Cart cart) {
+        synchronized (cart) {
```

Comment:
Remove synchronized from this private method. It creates additional performance costs.
Simply put recalculateCart in synchronized blocks of public methods from where it's called